pkg/adapter/awscognitouserpoolsource/adapter_test.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright (c) 2019-2020 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awscognitouserpoolsource
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/aws/aws-sdk-go/service/cognitoidentityprovider"
+	"github.com/aws/aws-sdk-go/service/cognitoidentityprovider/cognitoidentityprovideriface"
+
+	adaptertest "knative.dev/eventing/pkg/adapter/v2/test"
+	loggingtesting "knative.dev/pkg/logging/testing"
+)
+
+type mockedCognitoUserPoolClient struct {
+	cognitoidentityprovideriface.CognitoIdentityProviderAPI
+	listUsersOutput []cognitoidentityprovider.ListUsersOutput
+}
+
+var counter int
+
+func (m mockedCognitoUserPoolClient) ListUsers(in *cognitoidentityprovider.ListUsersInput) (*cognitoidentityprovider.ListUsersOutput, error) {
+	// return both sets of "UserType" sequentially
+	if counter < len(m.listUsersOutput) {
+		defer func() { counter++ }()
+		return &m.listUsersOutput[counter], nil
+	}
+	return &m.listUsersOutput[0], nil
+}
+
+func TestStart(t *testing.T) {
+	pollInterval = 1
+
+	ceClient := adaptertest.NewTestClient()
+
+	user1 := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: func() *time.Time {
+			now := time.Now()
+			return &now
+		}(),
+	}
+	user2 := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: func() *time.Time {
+			now := time.Now().Add(time.Second)
+			return &now
+		}(),
+	}
+
+	a := &adapter{
+		userPoolID: "userpool/fooPool",
+		logger:     loggingtesting.TestLogger(t),
+		ceClient:   ceClient,
+		cgnIdentityClient: mockedCognitoUserPoolClient{
+			listUsersOutput: []cognitoidentityprovider.ListUsersOutput{
+				{
+					Users: []*cognitoidentityprovider.UserType{
+						&user1,
+					},
+				}, {
+					Users: []*cognitoidentityprovider.UserType{
+						&user1,
+						&user2,
+					},
+				},
+			},
+		},
+	}
+
+	go func() {
+		err := a.Start(nil)
+		assert.NoError(t, err)
+	}()
+	// Wait while first UserType array processed
+	time.Sleep(3 * time.Second)
```

Comment:
I'm sure we can do better than that by receiving a notification on a channel once the processing completes. I can check tomorrow.

Reply:
Maybe we should just test start/stop separately, and just test the handler here, unless you just want to provide a single high level test. That's what I did in the EventBridge target.

Reply:
agree, test looks ugly. I'll work on improvement tomorrow

Reply:
@antoineco it looks like you don't use `sharedmain` package for adapter in EventBridge, right? 
I've tried to move code from Start() to make it a bit more testable. It ain't much, but it's honest work :man_farmer:

Reply:
That's right, I don't use the shared main but it should be possible to test `Start()` in the same way by simply initializing you adapter struct. I'll take another look.